core/server/worker/src/test/java/alluxio/worker/grpc/ShortCircuitBlockReadHandlerTest.java

```diff
@@ -0,0 +1,311 @@
+        mock(UfsManager.class), new AtomicReference<>(1L));
+
+    mTestHandler = new ShortCircuitBlockReadHandler(mBlockStore, mResponseObserver);
+  }
+
+  @Test
+  public void accessNonExistentBlock() {
+    // we started with empty directories without any block
+    long blockId = 1L;
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+    mTestHandler.onNext(request);
+
+    // check that we get a proper error and no response
+    assertTrue(mResponseObserver.getResponses().isEmpty());
+    assertFalse(mResponseObserver.isCompleted());
+    Throwable t = mResponseObserver.getError();
+    assertTrue(t != null && t.getMessage().contains("BlockMeta not found"));
+  }
+
+  @Test
+  public void accessBlockNoPromote() throws Exception {
+    accessBlock(false);
+  }
+
+  @Test
+  public void accessBlockPromote() throws Exception {
+    accessBlock(true);
+  }
+
+  @Test
+  @Ignore("This test currently takes too long")
+  public void accessBlockPinned() throws Exception {
+    long blockId = 2L;
+    createLocalBlock(
+        1L, blockId, 0);
+
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+    mTestHandler.onNext(request);
+
+    // block 2 should be pinned now
+    // and an attempt to remove the block from local store should fail
+    //todo(yangchen): TieredBlockStore's REMOVE_BLOCK_TIMEOUT_MS is too long for testing
+    assertThrows(Exception.class, () -> mBlockStore.removeBlock(3L, blockId));
+  }
+
+  @Test(timeout = 5000)
+  public void unpinBlockOnError() throws Exception {
+    long blockId = 2L;
+    createLocalBlock(
+        1L, blockId, 0);
+
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+    mTestHandler.onNext(request);
+    mTestHandler.onError(new RuntimeException());
+
+    // block should be unpinned and another session can remove it
+    mBlockStore.removeBlock(3L, blockId);
+  }
+
+  @Test(timeout = 5000)
+  public void errorReAccessingBlock() throws Exception {
+    long blockId = 2L;
+    createLocalBlock(
+        1L, blockId, 0);
+
+    OpenLocalBlockRequest request = createRequest(blockId, false);
+
+    // first call should be ok
+    mTestHandler.onNext(request);
+    assertNull(mResponseObserver.getError());
+
+    // second call should cause an error
+    mTestHandler.onNext(request);
+
+    assertFalse(mResponseObserver.isCompleted());
+    assertNotNull(mResponseObserver.getError());
+
+    // the block should be unpinned
+    mBlockStore.removeBlock(3L, blockId);
+  }
+
+  @After
+  public void after() throws Exception {
+    mCloser.close();
+  }
+
+  private OpenLocalBlockRequest createRequest(long blockId, boolean promote) {
+    return OpenLocalBlockRequest.newBuilder().setBlockId(blockId).setPromote(promote).build();
+  }
+
+  private void accessBlock(boolean promote) throws Exception {
+    // create a block in second tier(tier1) so that promotion will move block
+    long sessionId = 1L;
+    long blockId = 2L;
```

Comment:
These values are repeated multiple times in different tests. Make them class constants.